### PR TITLE
perf(MMU): discard CallStackReader

### LINKED ISSUE #689: `MMU` should read less in exoModule and use less the CallStackReader
- Hub could give directly sourceRam / targetRam. One less call to callStackReader if the src / tgt is the currentFrame
- for exoBytes, `MMU` could get it from the ram when it's an ram to exo instructions. For `exoToRamStransplant` too when copying precompile result, as it's already written in the tgt ram fictitious context number

### PR DESCRIPTION
**Pull request by letypequividelespoubelles**
This PR is a clean-up of the MMU. As MMU is not triggered during execution anymore, but is given by the HUB the relevant memory infos in MmuCall, we can discard the callStackReader. 
I introduce a better structure for MMIO's instructions to delete code duplicate too.

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/CodeCopy.java [hunk 1]
```diff
@@ -38,7 +38,8 @@ public CodeCopy(final Hub hub) {
     this.hub = hub;
     this.contract = hub.currentFrame().metadata();
 
-    this.targetId(hub.currentFrame().contextNumber())
+    this.exoBytes(Optional.of(hub.romLex().getCodeByMetadata(contract)))
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/CodeCopy.java:41]: Here `exoBytes` is made to contain (nothing or) the currently executing bytecode, correct ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create.java [hunk 2]
```diff
@@ -45,6 +45,7 @@ public Create(final Hub hub) {
                 hub.currentFrame()
                     .frame()
                     .shadowReadMemory(0, hub.currentFrame().frame().memoryByteSize())))
+        .exoBytes(Optional.of(hub.romLex().getCodeByMetadata(contract)))
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create.java:48]: Is `hub.romLex().getCodeByMetadata(contract)` already initialized to the `initCode` of the `CREATE` when this is executed ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create2.java [hunk 3]
```diff
@@ -45,6 +45,9 @@ public Create2(final Hub hub, final boolean failedCreate) {
                 hub.currentFrame()
                     .frame()
                     .shadowReadMemory(0, hub.currentFrame().frame().memoryByteSize())))
+        .exoBytes(
+            Optional.of(
+                hub.romLex().getCodeByMetadata(contract))) // TODO ? What about for failed CREATE2 ?
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create2.java]: For the failed CREATE2 we still need to send stuff to SHAKIRA. But not to ROM.

Comment (OlivierBBB) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create2.java]: I'm not sure how this works here but in any case you will need to activate both SHAKIRA and (potentially) ROM.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/instructions/LimbToRamOneTarget.java [hunk 4]
```diff
@@ -19,31 +19,19 @@
...
-    final MmuToMmioInstruction mmuToMmioInstruction =
-        mmuData.mmuToMmioInstructions().get(instructionNumber);
-
-    MmioData mmioData =
-        new MmioData(
-            mmuData.hubToMmuValues(),
-            mmuToMmioConstantValues,
-            mmuToMmioInstruction,
-            mmuData.exoSumDecoder());
+    final MmioData mmioData = super.execute();
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/instructions/LimbToRamOneTarget.java:34]: This is like in the precompile section I take it ? We have a generic method that runs initially and a more refined method that runs in the respective constructor ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MmuOperation.java [hunk 5]
```diff
@@ -91,14 +89,18 @@ protected int computeLineCount() {
   }
 
   public int computeMmioLineCount() {
-    int sum = 0;
+    if (mmioLineCount != -1) {
+      return mmioLineCount;
+    }
+
+    mmioLineCount = 0;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MmuOperation.java:96]: Was this method being called several times ? What explains this precaution ?

**Review by OlivierBBB — APPROVED** [review]

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/instructions/LimbToRamOneTarget.java (hunk 4 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/instructions/LimbToRamOneTarget.java:34]: Yes

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/CodeCopy.java (hunk 1 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/CodeCopy.java:41]: Correct

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MmuOperation.java (hunk 5 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MmuOperation.java:96]: Up to nb if tx in the batch, so ~200. We have the same mechanism for the hub line counting

**Comment by DavePearce** [comment]
@letypequividelespoubelles `go-corset` fails because of this:

```
[ERROR] missing input column 'hub.RELATIVE_BLOCK_NUMBER' in trace (AIR)
```

This column is not in the trace, but is defined in the constraints.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create.java (hunk 2 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create.java:48]: yes, just before :

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create2.java (hunk 3 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/opcode/Create2.java]: yes, you're absolutely right. Th

**Pull request merged**